python/dgl/nn/pytorch/glob.py

```diff
@@ -0,0 +1,673 @@
+        Compute multi-head self-attention.
+        """
+        feat_name = get_ndata_name(graph, self._feat_name)
+        query_name = get_ndata_name(graph, self._query_name)
+        key_name = get_ndata_name(graph, self._key_name)
+        value_name = get_ndata_name(graph, self._value_name)
+        score_name = get_ndata_name(graph, self._score_name)
+        att_name = get_ndata_name(graph, self._att_name)
+        out_name = get_ndata_name(graph, self._out_name)
+
+        # Copy q_feat and kv_feat to graph data frame
+        graph.nodes[q_nids].data[feat_name] = q_feat
+        graph.nodes[kv_nids].data[feat_name] = kv_feat
+
+        # Compute queries, keys and values.
+        graph.nodes[q_nids].data[query_name] =\
+            self.W_q(graph.nodes[q_nids].data[feat_name]).view(-1, self.num_heads, self.d_head)
+        graph.nodes[kv_nids].data[key_name] =\
+            self.W_k(graph.nodes[kv_nids].data[feat_name]).view(-1, self.num_heads, self.d_head)
+        graph.nodes[kv_nids].data[value_name] =\
+            self.W_v(graph.nodes[kv_nids].data[feat_name]).view(-1, self.num_heads, self.d_head)
+
+        # Free node features.
+        graph.ndata.pop(feat_name)
+
+        # Compute attention score.
+        graph.apply_edges(fn.u_mul_v(key_name, query_name, score_name))
+        e = graph.edata.pop(score_name).sum(dim=-1, keepdim=True) / np.sqrt(self.d_head)
+        graph.edata[att_name] = self.dropa(edge_softmax(graph, e))
+        graph.pull(q_nids,
+                   fn.u_mul_e(value_name, att_name, 'm'),
+                   fn.sum('m', out_name))
+        sa = self.W_o(graph.nodes[q_nids].data[out_name].view(-1, self.num_heads * self.d_head))
+        feat = self.norm_in(q_feat + sa)
+
+        # Free queries, keys, values, outputs and attention weights.
+        graph.ndata.pop(query_name)
+        graph.ndata.pop(key_name)
+        graph.ndata.pop(value_name)
+        graph.ndata.pop(out_name)
+        graph.edata.pop(att_name)
+
+        # Position-wise Feed Forward Network
+        feat = self.norm_inter(feat + self.ffn(feat))
+
+        return feat
+
+
+class SetAttentionBlock(nn.Module):
+    r""" SAB block mentioned in Set-Transformer paper."""
+    def __init__(self, d_model, num_heads, d_head, d_ff, dropouth=0., dropouta=0.):
+        super(SetAttentionBlock, self).__init__()
+        self.mha = MultiHeadAttention(d_model, num_heads, d_head, d_ff,
+                                      dropouth=dropouth, dropouta=dropouta)
+
+    def forward(self, feat, graph, sab_graph=None):
+        """
+        Compute a Set Attention Block.
+
+        Parameters
+        ----------
+        feat : torch.Tensor
+            The input feature
+        graph : DGLGraph
+            The graph.
+        sab_graph: DGLGraph or None
+            The graph to apply Message Passing on, set to None if not
+            specified.
+        """
+        if sab_graph is None:
+            sab_graph = _create_fully_connected_graph(graph)
```

Comment:
I wonder what's a better option here than creating an explicit graph. From the paper, the SAB blocks is roughly X\*X^T\*X. In the batched graph case, each graph can have different number of node features, does that mean we want to something like `segment_mm`? Wonder how other packages implement this.

Reply:
Also, can we apply these SetTransformer modules on edge features? It should be doable right?

Reply:
About the `sab_graph` argument. I agree that allowing this flexibility is good, but it might be hard to let user provide a such graph. At least, you need to give some descriptions (better with examples) about what's the requirement of the graph. If it's too complicated to explain, we could also just get rid of this argument.

Reply:
AFAIK no other GNN packages support Set Transformer, there are some 3rd-party implementations and they use padding.
Yes we can apply SetTransformer to edges, and so are other pooling algorithms. Though we could add an argument to specify `nodes`/`edges`, I think it's not necessary to do.

Reply:
I refactored the code and it no longer uses explicit graph.